### DISCUSSION TITLE
Efficient implementation of Stratonovich drift correction

### DISCUSSION THREAD

**Discussion by ylefay** [jax-ml/jax#16449] [Q&A] [answered]
Hello, thanks a lot for your work on this library!

I am trying to implement this function
$$b \mapsto \bigg(\sum_{j=1}^{n} \sum_{k=1}^{m} {b^{(j,k)}(x)\frac{\partial b^{(i,k)}(x)}{\partial x_j}}\bigg)_{i}, $$
which corresponds to the Ito to Stratonovich drift correction term for SDEs. Typically, $b$ is a function from $\mathbb{R}^n$ to $\mathbb{R}^{n\times m}$.

I wrote that
```python
def drift_correction(diffusion, x):
    return jnp.einsum('jk,ikj->i', diffusion(x), jax.jacfwd(diffusion)(x))
```
and wanted to know if it was possible to implement it  using ```jax.jvp```, ```jax.vjp``` or ```jax.linearize```.

WDYT?
Thanks.

Yvann

**Comment by anh-tong** [comment, marked as answer]
Hi, I think you approach is fair. As you mentioned, we can use `jax.jvp` to avoid the need for storing the matrix `jax.jacfwd(diffusion)(x)`, which would otherwise  require space of order $O(n \times m)$.  Below is an example for you case (see [this](https://github.com/patrick-kidger/diffrax/blob/a1003ef1c6899ad00890270b9f4a81857e757bde/diffrax/solver/milstein.py#L83C57-L83C57) in Diffrax)

```python
def drift_correction(diffusion, x):
    b = diffusion(x)
    _, corrected = jax.jvp(diffusion, (x, ), (b, ))
    return corrected
```
If you're interested in SDE, you may want to check out [Diffrax](https://github.com/patrick-kidger/diffrax/) which is designed for solving differential equations in JAX.
